### ISSUE TITLE
Attribute data incosistency in Draco payload when onlyChangedAttrs flag enabled in Draco subscription

### ISSUE THREAD

**Issue by achandana** [telefonicaid/fiware-orion#3666]
I am using "onlyChangedAttrs" in my Draco subscription.
 My draco subscription as below :

> curl -iX POST 'http://localhost:1026/v2/subscriptions?options=skipInitialNotification'  
>  -H 'Content-Type: application/json' 
>  -H 'fiware-service: test' 
>  -H 'fiware-servicepath: /test/iot/pf' 
>   -d '{   
>  "description": "Notify Draco of all context changes", 
>    "subject": {
>     "entities": [   
>    {      
>   "idPattern": ".*" 
>   } 
>  ]  
>  }, 
> "notification": { 
>  "http": {  
> "url": "http://localhost:5050/v2/notify"   
>  },  
> "onlyChangedAttrs":true 
>  }

Device registration paylod : 
{  
"devices": [  
{  
"device_id": "WSensor-3",  
"entity_name": "urn:ngsi-ld:SENSOR:WSensor-3",  
"entity_type": "SENSOR",  
"attributes": [  
        { "object_id": "temperature", "name":"temperature", "type":"Integer"  
        },
        { "object_id": "airpressure", "name":"temperature", "type":"Integer"  
        },
        { "object_id": "humidity", "name":"humidity", "type":"Integer"  
        },  
        { "object_id": "pressure", "name":"pressure", "type":"Integer"  
        },  
        { "object_id": "thing_id", "name":"thing_id", "type":"text"},  
        { "object_id": "thing_time", "name":"thing_time", "type":"text"}  
        ]  
}  
]  
}

When I send the measurement to device, in draco i am not getting thing_id and thing_time in some of the draco payloads.

Please find the below logs I have extracted.

Measurement payload :
   {"thing_id":"ABCD","thing_time":"2020-06-08T17:38:35.629+5:30","temperature":417}
{"thing_id":"ABCD","thing_time":"2020-06-08T17:38:35.629+5:30","airpressure":992.59,"humidity":1062.98}

{"thing_id":"ABCD","thing_time":"2020-06-08T17:38:35.629+5:30","temperature":650}
 {"thing_id":"ABCD","thing_time":"2020-06-08T17:38:35.629+5:30","airpressure":1065.14,"humidity":955.81}

Metadata:
{ "_id" : { "id" : "urn:ngsi-ld:SENSOR:WSensor-3", "type" : "SENSOR", "servicePath" : "/test/iot/pf" }, "attrNames" : [ "temperature", "humidity", "pressure", "thing_id", "thing_time", "TimeInstant" ], "attrs" : { "temperature" : { "value" : 992.59, "type" : "Integer", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373757 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373757 }, "humidity" : { "value" : 1062.98, "type" : "Integer", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373757 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373757 }, "pressure" : { "type" : "Integer", "creDate" : 1592314538, "modDate" : 1592314538, "value" : " ", "mdNames" : [ ] }, "thing_id" : { "value" : "ABCD", "type" : "text", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373757 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373757 }, "thing_time" : { "value" : "2020-06-08T17:38:35.629+5:30", "type" : "text", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373757 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373757 }, "TimeInstant" : { "value" : 1592373757, "type" : "DateTime", "mdNames" : [ ], "creDate" : 1592314538, "modDate" : 1592373757 } }, "creDate" : 1592314538, "modDate" : 1592373757, "lastCorrelator" : "253f36f4-b060-11ea-8e06-0242ac120104" }

{ "_id" : { "id" : "urn:ngsi-ld:SENSOR:WSensor-3", "type" : "SENSOR", "servicePath" : "/test/iot/pf" }, "attrNames" : [ "temperature", "humidity", "pressure", "thing_id", "thing_time", "TimeInstant" ], "attrs" : { "temperature" : { "value" : 1065.14, "type" : "Integer", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373817 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373817 }, "humidity" : { "value" : 955.81, "type" : "Integer", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373817 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373817 }, "pressure" : { "type" : "Integer", "creDate" : 1592314538, "modDate" : 1592314538, "value" : " ", "mdNames" : [ ] }, "thing_id" : { "value" : "ABCD", "type" : "text", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373817 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373817 }, "thing_time" : { "value" : "2020-06-08T17:38:35.629+5:30", "type" : "text", "md" : { "TimeInstant" : { "type" : "DateTime", "value" : 1592373817 } }, "mdNames" : [ "TimeInstant" ], "creDate" : 1592314538, "modDate" : 1592373817 }, "TimeInstant" : { "value" : 1592373817, "type" : "DateTime", "mdNames" : [ ], "creDate" : 1592314538, "modDate" : 1592373817 } }, "creDate" : 1592314538, "modDate" : 1592373817, "lastCorrelator" : "4918dc06-b060-11ea-a1b4-0242ac120104" }

Draco Payload :-

{"subscriptionId":"5ee8a9d51c90b6a7f3dc6437","data":[{"id":"urn:ngsi-ld:SENSOR:WSensor-3","type":"SENSOR","temperature":{"type":"Integer","value":992.59,"metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:02:37.00Z"}}},"humidity":{"type":"Integer","value":1062.98,"metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:02:37.00Z"}}}}]}

{"subscriptionId":"5ee8a9d51c90b6a7f3dc6437","data":[{"id":"urn:ngsi-ld:SENSOR:WSensor-3","type":"SENSOR"**,"thing_id":{"type":"text","value":"ABCD","metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:02:37.00Z"}}},"thing_time":{"type":"text","value":"2020-06-08T17:38:35.629+5:30","metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:02:37.00Z"}}**},"temperature":{"type":"Integer","value":417,"metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:02:37.00Z"}}},"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:02:37.00Z","metadata":{}}}]}

{"subscriptionId":"5ee8a9d51c90b6a7f3dc6437","data":[{"id":"urn:ngsi-ld:SENSOR:WSensor-3","type":"SENSOR","temperature":{"type":"Integer","value":1065.14,"metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:03:37.00Z"}}},"humidity":{"type":"Integer","value":955.81,"metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:03:37.00Z"}}}}]}

{"subscriptionId":"5ee8a9d51c90b6a7f3dc6437","data":[{"id":"urn:ngsi-ld:SENSOR:WSensor-3","type":"SENSOR",**"thing_id":{"type":"text","value":"ABCD","metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:03:37.00Z"}}},"thing_time":{"type":"text","value":"2020-06-08T17:38:35.629+5:30","metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:03:37.00Z"}}},**"temperature":{"type":"Integer","value":650,"metadata":{"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:03:37.00Z"}}},"TimeInstant":{"type":"DateTime","value":"2020-06-17T06:03:37.00Z","metadata":{}}}]}

In above Draco payloads, thing_id and thing_name are not coming in first and third payload. I have highlighted in second and fourth  payload .
What would be the the reason.

**Comment by onrao** [comment]
@fgalan  requesting your quick look into the issue we observed above as said by @achandana  requesting for suggestion, if  it is a known issue and it has got fixed  then the required configuration or any alternate fix for this issue

**Comment by fgalan** [comment]
It would be great if you could formulate the problem in terms of Orion Context Broker (note this is fiware-orion repository ;). I mean, simplify the case removing Draco and IOTA and describe it in in terms of Orion APIs interaction. That way it would be much easier to debug.

In particular, what I need is:

1. The subscripiton you have done (already provided above, great!)
2. An example of update received at Orion Context Broker which triggers the subscription (hint: you can look for that information in IOTA/Orion logs or capture traffic)
3. The notifications send by Orion as consequence of such update

In addition, please provide JSON fragments in pretty-printed form (tools like http://jsonlint.com may help) and using the gittub code style like below:

```json
{
   "foo": "This is just an example on how JSON is formated in github",
   "bar": 42
}
` ``

Finally, which version of Orion are you using? `GET /version` on Orion will provide such information.

Thanks!

**Comment by onrao** [comment]
Hi @fgalan Thank you for your quick attension. 
version of the Orion  is as follows,
{
  "orion": {
    "version": "2.3.0",
    "uptime": "30 d, 16 h, 37 m, 2 s",
    "git_hash": "764f44bff1e73f819d4e0ac52e878272c375d322",
    "compile_time": "Tue Nov 5 09:38:37 UTC 2019",
    "compiled_by": "root",
    "compiled_in": "38ab37448d3a",
    "release_date": "Tue Nov 5 09:38:37 UTC 2019",
    "doc": "https://fiware-orion.rtfd.io/en/2.3.0/"
  }
}

We had identified that our issue is similar to the issue # 3412 - support for milliseconds in timestamp attributes (Datetime type) #3412

We had followed the suggested approach of changing the datetype in IoT Agent Node lib 
TIMESTAMP_TYPE_NGSI2: 'DateTime' 
to 
TIMESTAMP_TYPE_NGSI2: "DateTimeWithMilis"

We are now able to get in all orion notifcation's all the payload data for all the topics published by IoT Agent to Orion.

Requesting this solution can be fine tuned using the seprate constants based on the configuration without impacting the backward compatability in IoT Agent /IoT Agent Node Lib  and pushing it into the mainline.


    /**

     * When this flag is active, the IoTAgent will add the TimeInstant attribute to every entity created, as well

     * as a TimeInstant metadata to each attribute, with the current timestamp.

     */

    **timestampsec: false,
    timestampwithmillisec: true**

Requesting your quick support to make it enabled ASAP in the mainline of the IoT Agent.

**Comment by onrao** [comment]
@fgalan do we need to raise this request again in IoT Agent SOF or GitHub, as we didn't have much time to prove this fiware based solution in my current project. requesting your qucik support and action to get this into mainline.
Looking forward to have a positive reply.

**Comment by fgalan** [comment]
So, if I understand correctly... you have solved the issue "hacking" the IOTA Node Lib? Is my interpretation correct?

**Comment by onrao** [comment]
Hi @ fgalan yes we had solved the issue by hacking the IoT Agent Nodelib. We would like to have a clean solution to be pushed into IoT Agent Node lib. Would you please help us?

**Comment by fgalan** [comment]
Would be a solution to support milliseconds in Orion DateTime attributes? (issues #3412, #2670 and #432)

**Comment by onrao** [comment]
Yes @fgalan It would be a common solution for the above indicated issues# 3412,#2670 and #432

**Comment by fgalan** [comment]
As far as I know, @kzangeli is working on milliseconds support in DateTime attributes in Orion and the work is pretty advance. Maybe @kzangeli himself can tell about this.

**Comment by fgalan** [comment]
> 
> 
> As far as I know, @kzangeli is working on milliseconds support in DateTime attributes in Orion and the work is pretty advance. Maybe @kzangeli himself can tell about this.

This is the work done by @kzangeli on this: https://github.com/telefonicaid/fiware-orion/pull/3668

@onrao @achandana : It would be great if you could test the fix. The simpler way would be to build a docker container based on the PR's branch. If you do so, your feedback about the test would be highly valuable. Thanks in avance!

**Comment by onrao** [comment]
@fgalan yes we can support on i.please let us know how we can enable it. Thanks

**Comment by fgalan** [comment]
PR #3673 (which is the sum of PRs #3668 and #3672) implementes milliseconds functionality for DateTime attributes and metadata into master.

> @fgalan yes we can support on i.please let us know how we can enable it. Thanks

If you use docker-based deployment it's just a matter of pull again Orion Context Broker container `:latest` tag and test again. If you do it, please tell us about the result of your tets. Thanks!

**Comment by fgalan** [comment]
Milliseconds support has been implemented for Orion 2.5.0 and, except if somebody tells otherwise, it works as expected. Closing issue.

**Comment by gsatheeshkumar** [comment]
@fgalan 

we have upgraded the Orion version (3.4.0) we have observed that millisecond-level tracking is not available in MongoDB.

as per your suggestions, the millisecond's support for DateTime attributes and metadata has been implemented. The feature will be available in the next Orion version (2.5.0).

but it is not supporting the latest version.

Can you please let us know how to handle these millisecond changes in the Orion subscription?

Appreciate your help!!

**Comment by fgalan** [comment]
Not sure what do you mean by "it is not supporting the latest version"...

Could you provide more detail on the sequence of steps (i.e. request to Orion context broker) you have done to reach that conclusion, please?
